specs/capri/Monitoring Design.md

```diff
@@ -0,0 +1,723 @@
+---
+title: "# Telemetry Design"
+---
+
+**Authors:** [Xing Yang](<https://github.com/xing-yang>), [Sanil Kumar D],
+Prakash R, Najmudheen CT
+
+This documentation serves as the design spec for monitoring and alerting
+
+## Background
+------------
+
+Currently there is no mechanism for monitoring in OpenSDS. Volume metrics will
+be introduced as the first step to support monitoring.
+
+## Objectives
+------------
+
+Here are the objectives for this design.
+
+### Goals
+------------
+
+* Retrieve volume level metrics from various storage backends.
+
+* Retrieve pool or backend level metrics from backends.
+
+* Retrieve metrics to measure storage operations.
+
+* Register metrics with Prometheus.
+
+* Provide APIs to expose metrics.
+
+* Display metrics in OpenSDS dashboard, integrating with Grafana.
+
+### Non Goals
+------------
+
+* Monitoring of Dock can be considered for the next phase.
+
+* Host information won’t be collected as new metrics, but we can get them from
+Prometheus node-exporter.
+
+* Network information won’t be collected here, but we can possibly get it from
+other tools
+
+* Monitoring of Gelato and multi-cloud
+
+## Architecture Diagram
+------------
+
+Components in the following diagram are shown in different colors. Boxes in blue
+represent components directly related to Telemetry. Components in green are part
+of Anomaly Detection module that consumes metrics from Telemetry. Only part of
+Hotpot is shown here for illustration purpose, but we do intent to collect
+metrics from other modules including Gelato as well (in a future release).
+
+![](media/dab0a7845ae69cccc46167cfbe089796.png)
+
+Note: Volume Metrics Exporter should be on both controller node and Dock node.
+This way a storage backend with only Cinder driver can implement only the metric
+volume interfaces and it can be used next to the Cinder driver.
+
+We can leverage existing exporters for Ceph backend.
+
+-   Ceph exporter available in Prometheus:
+    <https://github.com/digitalocean/ceph_exporter>
+
+-   Also got radosgw exporter:
+    <https://github.com/blemmenes/radosgw_usage_exporter>
+
+*A Detailed Diagram*
+
+![](media/be7e84d9043ecad275a20996ac0b855f.jpg)
+
+OpenSDS will provide Metrics REST APIs.
+
+The ML module of OpenSDS or a 3rd-party consumer of the OpenSDS Metrics module
+will make a POST request, asking for metrics, using OpenSDS Metrics REST API,
+possibly running some cron jobs. The request will be sent to the OpenSDS Metrics
+Controller, and then to the Dock.
+
+On the Dock there will be a Collector which collects metrics from Metric
```

Comment:
Yes, we will run a goroutine to call the driver API to collect metrics. Collected metric will be then pushed to Prometheus gateway.

Reply:
If so, u need to define an interface so that every driver could inherit it, are u going to define a new interface or just define these methods in VolumeDriver interface?

Reply:
We have defined a new interface for metrics, which can be implemented in the MetricDriver. Updated the document with the same.

Reply:
Cool